### DISCUSSION TITLE
Query parameters being lost when using Next.js with custom server

### DISCUSSION THREAD

**Discussion by MrAvantiC** [vercel/next.js#21185] [Help]
Hi there!

# Bug report

## Describe the bug

When following a certain navigation pattern in a Next.js app with a custom server, query parameters get lost - or not parsed properly, I'm not sure - in `getInitialProps`.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

I made a repo to reproduce the behaviour available here:
https://github.com/MrAvantiC/next-custom-server-query-param-issue

1. Go to the repo above, clone, install dependencies and start dev mode
2. Visit http://localhost:3000/ and observe both the terminal log as well as the browser console log.
3. Click on button "navigate with query param" which triggers a navigation using `Router.push`, adding a query parameter to the URL.
You will see in your browser-log that the `getInitialProps` in  `pages/index.js` is fired and `ctx.query` contains the added query parameter
4. Navigate to page "a" using the Link below the button.
5. Navigate back using the "browser back" functionality. 
You will again see the query parameter that was added before in the browser-log.

**So far, this is working as expected. Now to the weird part:** 

6. Hard-reload the page with the query parameter already added _or_ open a new tab and visit http://localhost:3000/?filter%5Bprice_to%5D=103.
In your server-log you will see the query parameter.
7. Navigate to page "a" using the Link below the button.
8. Navigate back using the "browser back" functionality. 
You will that now the query parameter is missing/incomplete in the browser-log

## Expected behavior

The expectation is that in the second navigation pattern the query parameter is identical to the first navigation pattern.

## System information

- OS: macOS
- Version of Next.js: 9.5.5
- Version of Node.js: 13.9.0

## Additional context

This is actually a breaking bug for us at the moment where we have E-Commerce customers that share deep-links to filtered product listings and if a user navigates to a detail page and then back, all the filters are lost. In some cases this even caused client-side errors.

What I also noticed is that if I comment out these lines in `server.js`:
```
  server.get('/', (req, res) => {
    return app.render(req, res, '/', req.query)
  })
```
...and restart the dev-mode, everything is working as expected. 
It seems to relate to using the routing of the custom server.

**Comment by MrAvantiC** [comment]
@timneutkens 
May I ask why you closed the issue?
Is something wrong or missing from the description?
As far as I can tell the issue still persists?!

**Reply by luflow** [reply]
@timneutkens We still have the same problem - should we reopen the issue or why was this moved to the discussion?

**Reply by Webbanditten** [reply]
I'm also running into this issue.
